### ISSUE TITLE
WebGL shared contexts and Webrender aren't synchronized

### ISSUE THREAD

**Issue by jdm** [servo/servo#21838]
With the following diff applied (to address #21796):
```diff
diff --git a/support/android/apk/servoview/src/main/java/com/mozilla/servoview/ServoView.java b/support/android/apk/servoview/src/main/java/com/mozilla/servoview/ServoView.java
index 0de1406847..5a3468e86a 100644
--- a/support/android/apk/servoview/src/main/java/com/mozilla/servoview/ServoView.java
+++ b/support/android/apk/servoview/src/main/java/com/mozilla/servoview/ServoView.java
@@ -198,10 +198,6 @@ public class ServoView extends GLSurfaceView
             mZoomFactor = 1;
         }

-        if (!zoomNecessary && !scrollNecessary && mAnimating) {
-            requestRender();
-        }
-
         if (mZooming || mScrolling || mAnimating) {
             Choreographer.getInstance().postFrameCallback(this);
         } else {
```
we end up with really smooth webGL rendering, except that parts of the view flicker in and out of existence. This is caused by using a shared GL context, meaning that the display list that webrender receives contains an external image type cotaining the GL context id. When webrender processes this display list, it ends up blitting the contents of the GL context in the right position. The ExternalImageHandler implementation in webgl_thread.rs is supposed to provide synchronization here between webrender and the webgl thread, but there's a problem - by the time webrender processes the display list, we have already returned from the layout operation that represents the point at which the webgl canvas is supposed to be presented to the compositor. There's a period of time in which script could have been executing new WebGL operations on this GL context, causing it to contain unexpected and transient data at the point when webrender locks the canvas and blits the context's contents. I think we need to prevent further GL operations once layout is complete until webrender is ready to read its contents.

**Comment by jdm** [comment]
Adding a `thread::sleep(Duration::from_millis(100))` to the end of layout made the de-synchronization problem go away, which increases my confidence in my analysis.

**Comment by jdm** [comment]
This has implications for preserveDrawingBuffer, too, since the context needs to be cleared after webrender is done with it and before any further GL drawing occurs.

**Comment by jdm** [comment]
The readback mode works fine because it obtains the pixels for the canvas during layout, so any subsequent modifications of the canvas aren't picked up when webrender processes the display list.

**Comment by asajeffrey** [comment]
This causes problems for some webVR/XR implementations too. Since we don't double buffer canvas textures, we end up having to do an extra blit on devices like ML1, where the rendering happens on the main thread, and the WebGL thread can't block on the main thread (since the main thread blocks on the WebGL thread, c.f. https://github.com/servo/servo/issues/22914). This extra blit would be avoided if canvas double-buffered. cc @Manishearth

**Comment by jdm** [comment]
This was fixed by #24482.
